### PR TITLE
"Upload from URL" option in the upload form

### PR DESCRIPTION
**Pull request by akhil-rana**
An option to upload the slide image from URL has been added. 
This feature works with [SildeLoader #28](https://github.com/camicroscope/SlideLoader/pull/28)

Note that the speed of upload depends on both the source and the destination.
ezgif-6-7e4b2cbb6db9

### CODE REVIEW

File: apps/table.html [hunk 1]
```diff
@@ -657,6 +687,38 @@ <h3 class="text-center h3 mb-0">Available Slides</h3>
...
+
+	function urlInputChange()
+	{
+		const urlInput = $('#urlInput');
+		const url = urlInput.val();
+
+		if (!isValidHttpUrl(url)) {
+			urlInput.addClass('is-invalid');
+			urlInput.parent().append(`<div class="invalid-feedback" id="inputUrl-feedback0">
+					Enter valid URL </div>`);
```
Comment (cjchirag7) [inline apps/table.html:707]:
```suggestion
if (urlInput.parent().children().length === 2) {
				urlInput.parent().append(`<div class="invalid-feedback" id="inputUrl-feedback0">
					Enter valid URL </div>`);
			}
```
It is better to check, whether there already exists an error message, and append only if needed. Currently, if the user clicks on 'Upload' button with invalid URL 6-7 times, the error message gets append 6-7 time , as shown in this **screenshot** -
Screenshot from 2020-04-10 11-14-21

File: apps/table.html [hunk 2]
```diff
@@ -83,9 +83,8 @@
 	</script>
 	<title>CaMicroscope Data Table</title>
 
-	<script src="https://code.jquery.com/jquery-3.4.1.slim.min.js"
-		integrity="sha384-J6qa4849blE2+poT4WnyKhv5vZF5SrPo0iEjwBvKU7imGFAV0wwj1yYfoRSJoZ+n"
-		crossorigin="anonymous"></script>
+	<script src="https://code.jquery.com/jquery-3.4.1.min.js"
+		></script>
```
[the remark this answers is not in the document]
Comment (akhil-rana) [reply apps/table.html:87]: I made a jquery ajax call (may not be necessary) but 'slim' version does not contain ajax functions.

**Comment by akhil-rana** [comment]
> Overall , it seems to be a cool feature. I would recommend to make the changes suggested in the comments.
> 
> Also, from the UI/UX point of view, I would suggest that instead of creating a Modal inside another Modal, it should be better to replace the `File Browse` row by `Enter file URL` row, when user clicks on 'Upload via URL', in the same modal. And now, an option must be shown , saying 'Upload file from system' , clicking on which will again replace the `Enter file URL` row by `File Browse` row. The need for this shall be clear in **this screencast**, when user tries to upload a file from URL, but then wants to upload the file by browsing, instead -
> bug
> 
> So, with this approach, also it must stop the upload via the first process, when the user shifts to the other process . For eg., when user is uploading the file through URL, and suddenly shifts to browsing from system, then the upload through URL must stop, without any wait , and vice-verca.

Oh yes, I missed that.
I can remove the file browsing input while the url is uploading. That should fix this

**Comment by cjchirag7** [comment]
> Oh yes, I missed that.
> I can remove the file browsing input while the url is uploading. That should fix this

@akhil-rana , I would suggest removing this modal and instead, replacing the `Browse field` by `Enter URL` field and replacing `upload using URL` by `upload from system` option.

**Comment by akhil-rana** [comment]
> > Oh yes, I missed that.
> > I can remove the file browsing input while the url is uploading. That should fix this
> 
> @akhil-rana , I would suggest removing this modal and instead, replacing the `Browse field` by `Enter URL` field and replacing `upload using URL` by `upload from system` option.

Okay. Nice. 
I'm on it. Thanks for all the suggestions.

File: apps/table.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (cjchirag7) [reply apps/table.html:87]: Ok , cool !

**Comment by akhil-rana** [comment]
ezgif-4-01dc0f6b2e41

File: apps/loader/chunked_upload.js [hunk 3]
```diff
@@ -170,3 +170,92 @@ function finishUpload() {
...
+  }
+  // Add columns
+  fnametr.insertCell(-1).innerHTML = `<input type=text class="form-control" name=filename id='filename0'
+    onchange=fileNameChange(); value='${fileName}'>`;
+  fileNameChange();
+  tokentr.insertCell(-1).innerHTML = `<input type=text class="form-control" onchange=hideCheckButton();hidePostButton();
+    disabled name=token id='token0'>`;
+  slidetr.insertCell(-1).innerHTML = `<input type=text class="form-control" name=slidename id='slidename0'>`;
+  filtertr.insertCell(-1).innerHTML = `<input type=text class="form-control" name=filter id='filter0'>`;
+  document.getElementById('token'+0).value = token;
```
[the remark this answers is not in the document]
Comment (cjchirag7) [reply apps/loader/chunked_upload.js:207]: @akhil-rana , please have a look at this also.

**Comment by akhil-rana** [comment]
> Looks fine! But I wonder if the loading spinner can also be shown when the file is being uploaded from the local system, as the upload mechanism seems to be the same there also.

It can be done but won't be visible for local hosting and small files though.

File: apps/loader/chunked_upload.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (akhil-rana) [reply apps/loader/chunked_upload.js:180]: I removed the common part and made a new function. You can take a look.

[the remark this answers is not in the document]
Comment (cjchirag7) [reply apps/loader/chunked_upload.js:180]: Cool !

**Comment by birm** [comment]
There's a semi-related case that may/may not be worth adding; existing files. This is mostly used when files are introduced through some unknown means (e.g. mounting an images dir with existing slides). We'd know the slidename, and we wouldn't have to upload/finish in that case, since the file would already be ready.
Does it make sense to add this onto this PR/app, or more towards the still-conceptual bulk loader, do you think?

**Comment by akhil-rana** [comment]
This 'upload from URL' feature just adds the file into the `/images/uploading/` temp directory and the rest of the work is completely done by the previous implementations of `finish uploads` and such.

Now, according to current workflow file with same name is not allowed to be shifted to `/images/` folder from `images/uploading/` folder. But the same files with different names can exist anyway.

Maybe we can add a feature like:
Giving an option to use the pre existing image file or replacing it with new file if same file name is encountered.

But I don't think it is a part of this.

**Pull request merged**